db/active_replicator_checkpointer.go

```diff
@@ -265,9 +274,26 @@ func (c *Checkpointer) _updateCheckpointLists() (safeSeq *SequenceID) {
 		base.TracefCtx(c.ctx, base.KeyReplicate, "checkpointer: _updateCheckpointLists removed seq %v from processedSeqs map", removeSeq)
 	}
 
-	// trim expectedSeqs list for all processed seqs
+	// trim expectedSeqs list from beginning up to first unprocessed seq
 	c.expectedSeqs = c.expectedSeqs[maxI+1:]
 
+	// if we have many remaining expectedSeqs, see if we can shrink the lists even more
+	// compact contiguous blocks of sequences by keeping only the last processed sequence in both lists
+	if len(c.expectedSeqs) > c.expectedSeqCompactionThreshold {
+		// start at the one before the end of the list (since we know we need to retain that one anyway, if it's processed)
+		for i := len(c.expectedSeqs) - 2; i >= 0; i-- {
+			current := c.expectedSeqs[i]
+			next := c.expectedSeqs[i+1]
+			_, processedCurrent := c.processedSeqs[current]
+			_, processedNext := c.processedSeqs[next]
+			if processedCurrent && processedNext {
+				// remove the current sequence from both sets, since we know we've also processed the next sequence and are able to checkpoint that
+				delete(c.processedSeqs, current)
+				c.expectedSeqs = append(c.expectedSeqs[:i], c.expectedSeqs[i+1:]...)
```

Comment:
If we had a large block of processed sequences, it would be nice to do this append just once, but I suppose the performance improvement doesn't merit the additional complexity/potential for bugs in this case?

Reply:
As discussed, this feels too complex to implement across multiple ranges for the benefits it'd give (given we're not growing the underlying array, it shouldn't be allocating/copying)